File: dlx.py
import numpy as np

class Node:
    """
    A class to represent the circular doubly linked list nodes. 
    """
    def __init__(self):
        self.left = self
        self.right = self
        self.up = self
        self.down = self
        self.column = self
        self.row_id = 0

class ColumnHeader(Node):
    """
    A class to represent the header column nodes to keep track of size and id,
    not to be treated as storing relevant data.

    Extends class Node.
    """
    def __init__(self, id):
        super().__init__()
        self.id = id
        self.size = 0

class DancingLinks:

    def __init__(self, cover_matrix):
        self.header = ColumnHeader("root")
        cover_matrix.insert(0, [0 for _ in range(len(cover_matrix[0]))])
        self.cover_matrix = cover_matrix;
        self.num_rows = len(self.cover_matrix)
        self.num_cols = len(self.cover_matrix[0])
        
        # dtype="object_" --> self.matrix will be populated by DLL nodes
        self.matrix = np.zeros((self.num_rows, self.num_cols), dtype="object_")

        self.partial_solution = [] # stacc
        self.result = []

    def generate_dll_mat(self):
        """
        Populates object attribute self.matrix with nodes

        Creates ColumnHeader nodes then populates the matrix with its vertical
        and horizontal links of all data nodes.
        """

        # Generate ColumnHeader nodes on first row
        for c in range(self.num_cols):
            col_node = ColumnHeader(c) 
            # Insert col_node to the left of the header node, update links
            col_node.left = self.header.left        # 0 <-> "root"
            col_node.right = self.header            # 0 <-> 1 <-> "root"
            # Update the link of the previous node to the left of the header
            self.header.left.right = col_node
            self.header.left = col_node

            # Insert the linked col_node into the matrix
            self.matrix[0][c] = col_node
        
        # Generate and link Nodes, populate the matrix.
        for r in range(1, self.num_rows):
            ref_node = None     # initialize ref_node for linking rows
            for c in range(self.num_cols):
                # Check the exact cover problem matrix 
                if self.cover_matrix[r][c] == 1:
                    # Create a new node, create its links
                    new_node = Node()
                    # Create vertical links
                    new_node.column = self.matrix[0][c] # c repr the col id
                    new_node.up = new_node.column.up                #    H1
                    new_node.down = new_node.column                 #    n1
                    new_node.column.up.down = new_node              # -> n2
                    new_node.column.up = new_node                   #    H1
                    
                    # Assign other properties to node
                    new_node.column.size += 1
                    new_node.row_id = r

                    # Create horizontal links
                    if ref_node == None: # in case of first node
                        ref_node = new_node
                    
                    new_node.left = ref_node.left #  n0 > new_node > ref
                    new_node.right = ref_node
                    ref_node.left.right = new_node
                    ref_node.left = new_node

                    self.matrix[r][c] = new_node


    def get_col_with_fewest_nodes(self):
        """
        Helper function to get the column header with the fewest nodes. Reduces
        the number of branches in the search.

        Returns:
        A ColumnHeader node with the fewest nodes.
        """

        col_with_fewest_nodes = head = self.header.right # move the head once
        
        # cycle check
        while head != self.header:
            if head.size < col_with_fewest_nodes.size:
                col_with_fewest_nodes = head

            head = head.right 
    
        return col_with_fewest_nodes

    def cover(self, target_node):
        """
        Covers a column. Traverses columns downwards and rows rightwards.

        Covering a column means finding all the 1s in the column and traversing
        their respective rows in order to remove the vertical links of all the
        1s in the row.

        Args:
            target_node: a Node object to cover the column of.
        """
        target_col = target_node.column

        # Un-link target_col
        target_col.left.right = target_col.right
        target_col.right.left = target_col.left

        # Iterate through the removed target_col's column (downwards)
        col_node = target_col.down

        while col_node != target_col: # cycle detection

            # Iterate through the row (rightwards)
            row_node = col_node.right
            while row_node != col_node: # cycle detection
                
                # Remove the links
                row_node.up.down = row_node.down
                row_node.down.up = row_node.up

                # Update the number of nodes the column header has
                self.matrix[0][row_node.column.id].size -= 1

                row_node = row_node.right # keep iterating row (rightwards)
            
            col_node = col_node.down # keep iterating col (downwards)

    def uncover(self, node):
        """
        Uncovers a column. Traverses columns upwards and rows leftwards.
        """
        
        target_col = node.column
        col_node = target_col.up

        # Iterate through column nodes (upwards)
        while col_node != target_col:

            # Iterate through row nodes (leftwards)
            row_node = col_node.left
            while row_node != col_node: 

                # Add the vertical links back for the row nodes
                row_node.up.down = row_node
                row_node.down.up = row_node

                # Update the number of nodes in the ColumnHeader node
                self.matrix[0][row_node.column.id].size += 1

                row_node = row_node.left
            
            col_node = col_node.up

        # Add back in the links to the now-uncovered column header
        target_col.left.right = target_col
        target_col.right.left = target_col

    def search(self, k=0):
        """
        Backtracking searcher to find solutions to the exact cover problem.

        Args:
            k: an int representing the depth of search. Start it at 0 lol
        
        Returns:
        The solution!!!!!!!!!!!!!!!!!!!!!!!!11
        """
        # Base case -- all columns have been covered
        if self.header.right == self.header:
            self.result.append(self.partial_solution[:])
            return
        

        col_header = self.get_col_with_fewest_nodes()

        self.cover(col_header)

        # Iterate through the covered column
        dummy_col = col_header.down
        while dummy_col != col_header:
            self.partial_solution.append(dummy_col.row_id)
    
            dummy_row_right = dummy_col.right

            
            # Iterate through the rows of the covered column's 1s & cover them
            while dummy_row_right != dummy_col:
                self.cover(dummy_row_right)
                dummy_row_right = dummy_row_right.right
            # Recurse -- committing to this partial solution
            self.search(k+1)

            # Backtrack
            self.partial_solution.pop() # pop from the stack

            col_header = dummy_col.column # <- I believe this line is redundant
            dummy_row_left = dummy_col.left
            while dummy_row_left != dummy_col:
                self.uncover(dummy_row_left)
                dummy_row_left = dummy_row_left.left
            
            dummy_col = dummy_col.down # move the dummy col node down
        
        self.uncover(col_header)
        return # no solutions

    def solve(self):
        self.generate_dll_mat()
        self.search()
        final = min(self.result, key=len)
        print(f"Our final result: {final}, all the solutions are: {self.result}")
        return len(final)
    
    
TEST_COVER_MATRIX = [
    [1, 0, 0, 1],
    [0, 1, 0, 1],
    [0, 0, 1, 0],
    [1, 0, 0, 0],
    [1, 0, 1, 0]
]

dlx = DancingLinks(TEST_COVER_MATRIX)
dlx.solve()

File: src/dlx.js
/**
 * Contains functionality for the Dancing Links Algorithm (DLX)
 */


/**
 * A class that represents a node in a circular doubly linked list
 */
class Node {
    constructor() {
        this.up = this;
        this.down = this;
        this.right = this;
        this.left = this;
        this.column = this;
        this.row_id = 0;
    }
}

/**
 * Class to represent the Column Header nodes.
 */
class ColumnHeader extends Node {
    constructor(id) {
        super();
        this.size = 0;
        this.id = id;
    }
}

export class DLX {
    /**
     * 
     * @param {Int[][]} coverMatrix 
     */
    constructor(coverMatrix) {
        this.header = new ColumnHeader("root");
        
        coverMatrix.unshift(Array.from( {length: coverMatrix[0].length}, () => 0));
        this.coverMatrix = coverMatrix;
        console.log(this.coverMatrix);
        this.numRows = coverMatrix.length;
        this.numCols = coverMatrix[0].length;
        this.matrix = Array.from({ length: this.numRows}, () =>
            Array.from({ length: this.numCols }, () => null)
        );
     
        
        this.results = new Array();
        this.partialSolutions = new Array();

        this.searchSteps = [];
        this.logs = [];

    }

    generateDLLMat() {
        /* Generate all column headers */
        for (let i = 0; i < this.numCols; i++) {
            let colHeader = new ColumnHeader(i);

            // Link the new ColumnHeader with its neighbors and vice versa
            colHeader.left = this.header.left;
            colHeader.right = this.header;

            this.header.left.right = colHeader;
            this.header.left = colHeader;

            this.matrix[0][i] = colHeader;
        }
        /* Iterate through all indices of this.matrix to make + link nodes*/
        for (let r = 1; r < this.numRows; r++) { // start at row after headers
            let refNode = null;
            
            for (let c = 0; c < this.numCols; c++) {
                if (this.coverMatrix[r][c] == 1) {

                    let newNode = new Node();
                    // Create vertical links
                    newNode.column = this.matrix[0][c];
                    newNode.up = newNode.column.up;
                    newNode.down = newNode.column;

                    newNode.column.up.down = newNode;
                    newNode.column.up = newNode;

                    // Update instance variables to reflect adding a new node
                    newNode.row_id = r;
                    newNode.column.size += 1;

                    // Create horizontal links
                    if (refNode == null) { // case where refNode is null, first iter
                        refNode = newNode;
                    }

                    newNode.left = refNode.left;
                    newNode.right = refNode;
                    refNode.left.right = newNode;
                    refNode.left = newNode;

                    this.matrix[r][c] = newNode;
                }
            }
        }
    }

    /**
     * Helper function to select the column with the fewest 1s.
     * @returns {ColumnHeader} column with the fewest 1s
     */
    selectColumn() {
        let selectedCol = this.matrix[0][0];
        for (let i = 1; i < this.numCols; i++) {
            if (this.matrix[0][i].size < selectedCol.size) {
                selectedCol = this.matrix[0][i];
            }
        }
        // console.log("Size of selected col: ", selectedCol.size); //FIXME:
        return selectedCol;
    }

    /**
     * Covers a column.
     * @param {Node} target the node whose column will be covered :O
     * @returns a dictionary with set values
     */
    cover(target, currDepth) {
        const affectedRows = [];

        // Cover the column - remove its links from the other ColHeaders
        let coveredCol = target.column;
        coveredCol.left.right = coveredCol.right;
        coveredCol.right.left = coveredCol.left;

        let row = coveredCol.down;
        while (row != coveredCol) {
            affectedRows.push(row.row_id);
            let rowRight = row.right;
            // Chop vertical links of rows
            while (rowRight != row) {
                rowRight.up.down = rowRight.down;
                rowRight.down.up = rowRight.up;
                this.matrix[0][rowRight.column.id].size -= 1;
                rowRight = rowRight.right;
            }
            
            row = row.down;
        }


        this.logs.push({
            action: 'cover',
            column: target.id,
            affectedRows,
            depth: currDepth,
            //matrixState: this.getMatrixSnapshot()
        });
    }
    /**
     * Uncovers a column.
     * @param {Node} target the node whose column will be uncovered :D
     */
    uncover(target, currDepth) {
        const affectedRows = [];
        let uncoveredCol = target.column;

        let row = uncoveredCol.up;
        while (row != uncoveredCol) {
            affectedRows.push(row.row_id)
            let rowLeft = row.left;
            while (rowLeft != row) {
                // Add vertical links back in
                rowLeft.up.down = rowLeft;
                rowLeft.down.up = rowLeft;
                this.matrix[0][rowLeft.column.id].size += 1;
                rowLeft = rowLeft.left;
            }

            row = row.up;
        }
        
        // Add links from neighbors to uncoveredCol back in
        uncoveredCol.left.right = uncoveredCol;
        uncoveredCol.right.left = uncoveredCol;

        this.logs.push({
            action: 'uncover',
            column: target.id,
            affectedRows,
            depth: currDepth,
            //matrixState: this.getMatrixSnapshot()
        });
    }
    /**
     * Backtracking algorithm to find a solution given the cover matrix.
     * @param {Int} k represents the depth of the backtracking search
     */
    search(k) {
        // if (k > 5) {
        //     return;
        // }
        if (this.header == this.header.right) {
            this.results.push([...this.partialSolutions]);
            return;
        }   

        // Select and cover the col with fewest 1s
        let selectedCol = this.selectColumn()
        console.log("Selected column:", selectedCol.id, "size:", selectedCol.size, "depth: ", k);
        this.cover(selectedCol, k);

        let col = selectedCol.down;
        while (col != selectedCol) {
            this.partialSolutions.push(col.row_id);

            console.log("I'm in column ", col.column.id);

            // Cover
            let rowR = col.right;
            while (rowR != col) {
                console.log("covering ", rowR.row_id, " ", rowR.column.id);
                this.cover(rowR, k);
                rowR = rowR.right; 
            }
            // Recurse
            this.search(k+1);

            // Backtrack
            this.partialSolutions.pop();
            
            // Uncover in the opposite direction
            let rowL = col.left;
            while (rowL != col) {
                console.log("uncovering ", rowL.row_id, " ", rowL.column.id);
                this.uncover(rowL, k);
                rowL = rowL.left;
            } 

            col = col.down;
        }
        this.uncover(selectedCol, k);
        return;
        
    }

    solve() {
        this.generateDLLMat();
        this.search(0);
        let solution = this.results.reduce((min, current) => 
            current.length < min.length ? current : min, this.results[0]);
        return solution;
    }
}

let TEST_MATRIX = [
    [1, 0, 0, 1],
    [1, 1, 0, 1],
    [0, 0, 1, 0],
    [1, 0, 0, 0],
    [1, 1, 1, 0]
]
// let dlx = new DLX(TEST_MATRIX);
// // dlx.generateDLLMat();

// let sol = dlx.solve();
// console.log(dlx.results);
